### ISSUE TITLE
Reverting a page to a specific version is not working correctly : SS 3.2

### ISSUE THREAD

**Issue by priyashantha** [silverstripe/silverstripe-framework#5040]
When we use 'Revert to this version' in History tab of a cms page, it creates a new version record for child Page_versions table but not in SiteTree_versions (Page extends SiteTree and when Page has db array). 

Then we can't do anymore reverts as we have orphaned record in Page_versions table. This happens only in SS 3.2 version but not in SS 3.1

**Comment by tractorcow** [comment]
Have you tested this in 3.2.x-dev branch, or only 3.2.1 stable?

**Comment by tractorcow** [comment]
Ok, it looks like this could be an issue.


```
$from->forceChange();
if($createNewVersion) {
	$latest = self::get_latest_version($baseClass, $this->owner->ID);
	$this->owner->Version = $latest->Version + 1;
} else {
	$from->migrateVersion($from->Version);
}
```

The `$this->owner->Version` should be `$from->Version`.

**Comment by priyashantha** [comment]
I tested this with 3.2.1. Please see my reply in PR #5041.

**Comment by priyashantha** [comment]
I think we have the issue in $manipulation array according to my scenario.

in augmentWite() function line number 663: if(empty($manipulation[$table]['fields']['Version']))
Here we have 2 tables (Page and SiteTree). Above condition is false only for SiteTree table as it contains Version field but not for Page table. I tested this against to SS3.1 installation and there we have Version field for both of Page and SiteTree tables. That's why there is no new version record in SS3.1 when rollbacking.

**Comment by tractorcow** [comment]
Putting a duplicate Version field on each table will cause side effects... you are essentially de-normalising the DB schema and introducing the risk of breakage in case that the duplicated Version field gets de-synchronised.

`$version = isset($manipulation[$table]['fields']['Version'])`

Let's just change this so the version is obtained from the base table, not the same table being written.

We'll also need to adjust the logic in augmentWriteVersioned.

**Comment by priyashantha** [comment]
Here I don't understand what should be the final solution. In 3.2 release, are we creating new version records in all `_versions` tables for the class ancestry? Currently with the check `empty($manipulation[$table]['fields']['Version'])` in `augmentWrite()`, we have allowed to create a new version records only in child class tables(i.e. Page) but not the `$baseDataClass` table(i.e. SiteTree) given that field `Version` is not there in child tables and the referenced version already in the `$baseDataClass_versions`table (eg: when roll back)

Should that check not be like `empty($manipulation[$baseDataClass]['fields']['Version'])` because there is no point of checking field `Version` in each table as it'd be true only for `$baseDataClass`? If that the case we just update the version in stage table when roll backing but no new version record. 

If we're supposed to create a new version when roll backing what should be the scenario?

**Comment by tractorcow** [comment]
> In 3.2 release, are we creating new version records in all _versions tables for the class ancestry?

Yes we should be.

> Should that check not be like empty($manipulation[$baseDataClass]['fields']['Version']) because there is no point of checking field Version in each table as it'd be true only for $baseDataClass?

Maybe... I'll need to test this out. As with all issues that I'd like the time for to work out, that's the one thIng i don't have. ;) I appreciate your time in addressing this.

The problem is that SiteTree/Page_versions tables MUST have the version, as it's a part of the composite primary key, for all levels. Conversely, the draft tables SiteTree/Page must NOT duplicate this field, since the current value for any record ID is ONLY a single value (the one in SiteTree.Version).

This is why it's important to check the baseTable Version value when writing, but also why we cannot simply write Versions to each table (since the field doesn't even exist on every table).


To break it down, this is what writes shouldlook like

Page ID = 4, Version 4 is rolled back to version 3.
Page is loaded in memory, rolled back to 3, but version updated to 5.

SiteTree is written with ID = 4, Version = 5, plus values from rolled back record
Page is written with ID = 4, plus values from rolled back records
SiteTree_versions is written with RecordID = 4, Version = 5 (new ID auto-incremented)
Page_versions is written with RecordID = 4, Version = 5 (New ID is auto-incremented)


Thus, the stage record can be found by joining SiteTree to Page on ID.
The version record for that version can be found by joining SiteTree_versions on Page_versions on RecordID AND Version. This is because _versions tables all include multiple records for each recordID, which is why the versions field is mandatory on all tables here (and only here).

**Comment by tractorcow** [comment]
@priyashantha  can you please test out https://github.com/silverstripe/silverstripe-framework/pull/5074 and tell me if this fixes your issue?

I've tested it locally and it seems to revert subclassed pages properly.

**Comment by priyashantha** [comment]
@tractorcow I also tested this locally and it's working fine. Can you please let me know when the release is ready for this fix? :)

**Comment by dhensby** [comment]
Looks like the fix won't make it in until 3.4... Is that right @tractorcow

**Comment by tractorcow** [comment]
No, it's getting released in 3.2.3. AFTER I fix the possible regression in it (re https://github.com/silverstripe-labs/silverstripe-fulltextsearch/pull/109)
